refactor: update Readme.md v1 (#601)
asyncapi/parser-js: README.md

<!-- tocstop -->

## Installation

```bash
npm install @asyncapi/parser
yarn add @asyncapi/parser
```

> The parser by default supports AsyncAPI Schema Format and JSON Schema Format for schemas. For additional formats, check [Custom message parsers](#custom-schema-parsers) section.

Reviewer: ```suggestion
The parser by default supports AsyncAPI Schema Format and JSON Schema Format for schemas. For additional formats, check [Custom schema parsers](#custom-schema-parsers) section.
```

### [Example using Avro schemas](#custom-schema-parsers)

Head over to [asyncapi/avro-schema-parser](https://www.github.com/asyncapi/avro-schema-parser) for more information.

### [Example using OpenAPI schemas](#custom-schema-parsers)

Head over to [asyncapi/openapi-schema-parser](https://www.github.com/asyncapi/openapi-schema-parser) for more information.

### [Example using RAML data types](#custom-schema-parsers)

Head over to [asyncapi/raml-dt-schema-parser](https://www.github.com/asyncapi/raml-dt-schema-parser) for more information.

### [Example with stringify and unstringify parsed document](#stringify)

```ts
import { Parser, stringify, unstringify } from '@asyncapi/parser';

const parser = new Parser();

const { document } = await parser.parse(`
  asyncapi: '2.4.0'
  info:
    title: Example AsyncAPI specification
    version: '0.1.0'
  channels:
    example-channel:
      subscribe:
        message:
          payload:
            type: object
            properties:
              exampleField:
                type: string
              exampleNumber:
                type: number
              exampleDate:
                type: string
                format: date-time

> `);

Reviewer: This example is not valid to Spectral because of this rule `asyncapi-operation-operationId`: https://github.com/stoplightio/spectral/blob/297531bbc6f26393d325a101b59f9e8b1f57e21e/packages/rulesets/src/asyncapi/index.ts#L230-L240

Not sure if we really want to keep that rule there. The `operationId` is not mandatory by the spec.

Author: I know about some "problems" with Spectral. Now I would leave that spec, because at the end `operationId` won't be mandatory on Spectral side, don't worry.

Reviewer: Do you mean that particular spectral rule will be gone? Could you bring more details, please?

Author: Changed will be severity from error to warning - then there will be information that it is a good pattern design. Most of the rules from Spectral were made 2 years ago by the Stoplight team itself and a lot of things are not up to date. This week I have a plan to go through all the rules and improve them.

Reviewer: > Changed will be severity from error to warning

That makes totally sense to me. I have few concerns regarding this but I'm not gonna add anything extra into this issue, but rather please ping me on the work you plan to do 🙏

Author: I will create PR with that in Spectral side, so maybe better will be if you will list everything bugs that you see, hmm?

Direct access to the parsed JSON document is always available through the `doc.json()` method.

> See API documentation for more example and full API reference information.

Author: One of the workflows is showing us errors regarding the link with in this paragraph and I thought that by removing it the workflow would go through but it didn't I would leave it as is, because this section needs to be corrected anyway.

In case you just want to check out the latest `bundle.js` without installing the package, we publish one on each GitHub release. You can find it under [this link to the latest release](https://github.com/asyncapi/parser-js/releases/latest/download/bundle.js).

## Custom schema parsers

> AsyncAPI doesn't enforce one schema format. We can have payload of your messages described with OpenAPI (3.0.0), Avro, etc. This parser by default parses only [AsyncAPI Schema Format](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#schemaObject) (superset of JSON Schema Format). We can extend it by creating a custom parser and registering it within the parser:

Reviewer: ```suggestion
AsyncAPI doesn't enforce one schema format. The payload of the messages can be described with OpenAPI (3.0.0), Avro, etc. This parser by default parses only [AsyncAPI Schema Format](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#schemaObject) (superset of JSON Schema Format). We can extend it by creating a custom parser and registering it within the parser:
```

1. Create custom parser module that exports three functions:

    - `validate` - function that validates (its syntax) used schema.
    - `parse` - function that parses the given schema to the [AsyncAPI Schema Format](https://github.com/asyncapi/spec/blob/master/spec/asyncapi.md#schemaObject).
    - `getMimeTypes` - function that returns the list of mime types that will be used as the `schemaFormat` property to determine the mime type of a given schema.

    Example:
    
    ```ts
    export default {
      validate(input) { ... },
      parse(input) { ... },
      getMimeTypes() {
        return [
          'application/vnd.custom.type;version=1.0.0',
          'application/vnd.custom.type+json;version=1.0.0',
        ]
      }
    }
    ```

2. Before parsing/validating an AsyncAPI document with a parser, register the additional custom schema parser:

    ```ts
    import { Parser } from '@asyncapi/parser';
    import myCustomSchemaParser from './my-custom-schema-parser';

    const parser = new Parser();
    parser.registerSchemaParser(myCustomSchemaParser);
    ```

> ### Official supported custom schema parsers

Reviewer: I suggest we move this section to be at the top of `Custom schema parsers` section, as it will be easier for the user to find out info about those common parsers.

Author: After all, the introduction to this section is short, why move it up by these few lines?

Reviewer: Because is what users will look for **always** when looking for schema parsers: "Is there already an official parser for <whatever>?"

Author: What can I say? if they are impatient then let them cancel their Netflix account 😆
